src/compose/images.ts

```diff
@@ -417,35 +434,75 @@ async function getImagesForCleanup(): Promise<string[]> {
 		.value();
 }
 
-export async function inspectByName(
-	imageName: string,
-): Promise<Docker.ImageInspectInfo> {
-	try {
-		const image = await docker.getImage(imageName);
-		return await image.inspect();
-	} catch (e) {
-		if (NotFoundError(e)) {
-			const digest = imageName.split('@')[1];
-			let imagesFromDb: Image[];
-			if (digest != null) {
-				imagesFromDb = await db
-					.models('image')
-					.where('name', 'like', `%@${digest}`);
-			} else {
-				imagesFromDb = await db
-					.models('image')
-					.where({ name: imageName })
-					.select();
-			}
+export async function inspectByName(imageName: string) {
+	// Fail fast
+	assert(!!imageName, 'image name to inspect cannot be null');
 
-			for (const image of imagesFromDb) {
-				if (image.dockerImageId != null) {
-					return await docker.getImage(image.dockerImageId).inspect();
-				}
-			}
-		}
-		throw e;
-	}
+	// Get image by the full image URI. This will only work for regular pulls
+	// and old style images `repo:tag`
+	const inspectByURI = async () => await docker.getImage(imageName).inspect();
+
+	const {
+		registry,
+		imageName: name,
+		tagName,
+		digest,
+	} = dockerUtils.getRegistryAndName(imageName);
+
+	// Look for an image in the engine with registry/image as reference (tag)
+	// for images with deltas this should return unless there is some inconsistency
+	// and the tag was deleted
+	const repo = [registry, name].filter((s) => !!s).join('/');
+	const reference = [repo, tagName].filter((s) => !!s).join(':');
+
+	const inspectByReference = async () =>
+		await docker
+			.listImages({
+				digests: true,
+				filters: { reference: [reference] },
+			})
+			.then(([img]) =>
+				!!img
+					? docker.getImage(img.Id).inspect()
+					: Promise.reject(
+							new StatusError(
+								404,
+								`Failed to find an image matching ${imageName}`,
+							),
```

Comment:
`NotFoundError` seems to cover the 404 case already

Reply:
`NotFoundError` is a checker function, not a subclass of error

Reply:
Oops! My mistake